js/web/lib/wasm/jsep/webgpu/ops/group-query-attention.ts

```diff
@@ -0,0 +1,308 @@
+    } else if (key.dims.length === 5) {
+      if (key.dims[2] !== attributes.numHeads || key.dims[3] !== 2 || key.dims[4] !== headSize) {
+        throw new Error('Expect "key" shape (batch_size, kv_sequence_length, num_heads, 2, head_size) for packed kv');
+      }
+      if (value) {
+        throw new Error('Expect "value" be none when "key" has packed kv format.');
+      }
+      qkvFormat = AttentionQkvFormat.qKvBSNHxBSN2H;
+      kvSequenceLength = key.dims[1];
+    } else {  // key_dims.size() == 4 (cross-attention with past_key)
+      if (key.dims[1] !== attributes.numHeads || key.dims[3] !== headSize) {
+        throw new Error('Expect "key" shape (batch_size, num_heads, kv_sequence_length, head_size) for past_key');
+      }
+
+      qkvFormat = AttentionQkvFormat.unknown;
+      kvSequenceLength = key.dims[2];
+    }
+  } else {  // packed QKV
+    if (query.dims.length !== 3 && query.dims.length !== 5) {
+      throw new Error('Input "query" is expected to have 3 or 5 dimensions when key is empty');
+    }
+    if (query.dims.length === 5 && (query.dims[2] !== attributes.numHeads || query.dims[3] !== 3)) {
+      throw new Error('Expect "query" shape (batch_size, kv_sequence_length, num_heads, 3, head_size) for packed kv');
+    }
+
+    qkvFormat = AttentionQkvFormat.qkvBSN3H;
+  }
+
+  const maskType: AttentionMaskType = AttentionMaskType.none;
+  let passPastInKv = false;
+  let vHiddenSize = hiddenSize;
+  if (value) {
+    if (value.dims.length !== 3 && value.dims.length !== 4) {
+      throw new Error('Input "value" is expected to have 3 or 4 dimensions');
+    }
+
+    if (query.dims[0] !== value.dims[0]) {
+      throw new Error('Input "query" and "value" shall have same dim 0 (batch_size)');
+    }
+
+    if (value.dims.length === 3) {
+      if (kvSequenceLength !== value.dims[1]) {
+        throw new Error('Input "key" and "value" shall have the same dim 1 (kv_sequence_length)');
+      }
+      vHiddenSize = value.dims[2];
+    } else {
+      if (kvSequenceLength !== value.dims[2]) {
+        throw new Error('Input "past_key" and "past_value" shall have the same dim 2 (kv_sequence_length)');
+      }
+      vHiddenSize = value.dims[1] * value.dims[3];
+      passPastInKv = true;
+    }
+  }
+  const totalSequenceLength = pastSequenceLength + kvSequenceLength;
+  const broadcastResPosBias = false;
+
+  return {
+    batchSize,
+    sequenceLength,
+    pastSequenceLength,
+    kvSequenceLength,
+    totalSequenceLength,
+    maxSequenceLength,
+    inputHiddenSize: 0,
+    hiddenSize,
+    vHiddenSize,
+    headSize,
+    vHeadSize: Math.floor(vHiddenSize / attributes.kvNumHeads!),
+    numHeads: attributes.numHeads,
+    kvNumHeads: attributes.kvNumHeads,
+    nReps: attributes.numHeads / attributes.kvNumHeads!,
+    pastPresentShareBuffer: false,
+    maskType,
+    scale: attributes.scale,
+    broadcastResPosBias,
+    passPastInKv,
+    qkvFormat,
+  };
+};
+
+const createConcatProgramInfo =
```

Comment:
Do we sill need this local implementation of Concat? Does concat.ts not work?

Reply:
In the concat [spec](https://onnx.ai/onnx/operators/onnx__Concat.html),  it says "All input tensors must have the same shape". 

However, in attention, I didn't see same shape limits on past & new. On the contrary, I find some examples which requires different  shapes or sizes: 
1.  ONNX Spec
```
key (optional) : T
Key with shape (batch_size, kv_sequence_length, kv_hidden_size)
value (optional) : T
Value with shape (batch_size, kv_sequence_length, kv_hidden_size)
past_key (optional) : T
past state key with support for format BNSH. When past_key uses same tensor as present_key(k-v cache), it is of 
length max_sequence_length... otherwise of length past_sequence_length.
past_value (optional) : T
past state value with support for format BNSH. When past_value uses same tensor as present_value(k-v cache), it is of
 length max_sequence_length... otherwise of length past_sequence_length.
```
The new Key & Value are of kv_sequence_length, while past Key & Value are of max_sequence_length or past_sequence_length.  This means the new and past may possibly be of different sequence length.

2. llama
In https://github.com/meta-llama/llama/blob/main/llama/model.py#L285,  The cache(self.cache_k) will keep growing untill it reaches max_seq_len. self.cache_k will be used as present key(in current iteration) and past key in next interation. Usually the xk & xv (new Key & new Value) should be of a fixed shape, while the shape of self.cache_k is keep growing.

According to my understanding, concat requires same shape inputs, so it doesn't meet all those requirements from attention.